### DISCUSSION TITLE
IPFS 日用优化指南

### DISCUSSION THREAD

**Discussion by giscus** [zu1k/blog#58] [Comments]
# IPFS 日用优化指南

IPFS 是新兴的技术，本文借 Planet 发布之机，为大家提供一些 IPFS 的日用优化指南，包括搭建本地网关、Pin优化、优化资源占用等。在文末，我附加了自己对 IPFS 的一些看法。

https://zu1k.com/posts/tutorials/p2p/ipfs-easy-use/#giscus

**Comment by livid** [comment]
这件事情最早开始吸引我，是 2021 年初 Brave 开始提供对 IPFS 的原生支持。于是我当时尝试了将之前用 Jekyll 搭建的博客放到 IPFS 上，把最初的坑踩了一遍。

然后就被 IPFS 种草了。

然后就有了 Planet，因为那些命令行操作确实太麻烦了。如果不简化，是不可能让普通用户能用起来的。

**Reply by zu1k** [reply]
说一下我的看法。

依我对IPFS的使用经验，我心目中的IPFS应用模式应该是B/S架构，以家庭中的IPFS网关为核心，拓展出重多上层应用。

因为IPFS的资源开销不小，一个IPFS网关应服务一整个家庭，并且同一局域网中部署多个IPFS节点其实体验不会更好。并且考虑到个人电脑难以保证在线率，在软路由中部署IPFS网关应该是更加合适的使用方式。

所以我认为Planet应当想办法支持外部网关，甚至做成 C/S 架构，客户端仅关注内容呈现。

**Reply by livid** [reply]
是的，比如做成一个 Synology 的 App，然后让 Planet 可以配置要用哪个 API 和 gateway。

**Reply by zu1k** [reply]
因为我没有苹果电脑，所以并没有实际体验过 Planet，不过我稍微浏览了一下代码，并且根据群中截图大概能理解 Planet 的工作流程。

我觉得 Planet 目前的工作流程并没有抓住 IPFS 的特点，需要优化。IPFS 中静态内容是核心，动态内容的实现很困难。所以 Planet 应当抓住动与静的尺度。

目前 Planet 发布内容会生成列表 JSON，并根据模板渲染出内容；修改内容时是原地修改，并重新渲染。这些操作都会导致内容的改变，从而访问要重新进行内容寻址，这是动态网站的思路。我理想中的方式应当是类似于 Git 的模式，内容一旦发布就不再进行任何修改，所有的修改应当以叠加 patch 的方式进行，这种模式下，用户不仅可以追踪修改历史，还可以在内容变化时仅需要获取 patch。

并且我认为渲染应该在内容被查看时即时的在客户端进行，而不是在内容发布时进行。这样做的好处是用户可以随意的切换主题和显示风格，而不需要原作者重新发布内容，作者修改主题也仅需要修改planet描述文件即可，无需重新渲染整个网站。主题也应当通过 IPFS 方式发布。

**Reply by zu1k** [reply]
在使用 Hugo 生成本站并发布在 IPFS的使用体验中，可以发现这些“静态”网站生成器在增加或者修改任何内容时，也完全是动态网站的思路，需要重新渲染整个站点，包括分类/列表/标签，甚至毫不关联的其他文章也因为需要显示上一篇/下一篇的链接而需要重新渲染。这在 IPFS 中就造成了大量内容改变，需要多个页面的重新寻址。

我曾经想自己设计一款适合IPFS的静态网站生成器，在发布或修改内容时，仅修改尽可能少的索引列表，一切新内容都通过叠加的方式进行，patch可以按照规模进行定期合并和再发布。主题要动态渲染，就例如浏览器端动态渲染markdown一般，用户可以给浏览的网站随意切换自己喜欢的主题

**Reply by zu1k** [reply]
但是自己没有动力来实现，把这个期望保留，看好 Planet

**Reply by snowyu** [reply]
更简单的思路是在IPFS上存放markdown源码,渲染在浏览器上完成.这样还便于多人协作修改,把文档变成开源GIT项目

而PIN就是fork,然后再实现 pr

**Reply by livid** [reply]
Planet 确实把 MD 源码放上 IPFS 了：

https://olivida.eth.limo/planet.json

**Reply by snowyu** [reply]
这种json应该不算,人很难阅读比较,也无法merge

**Reply by zu1k** [reply]
可以像我这样 https://zu1k.com/posts/tutorials/p2p/ipfs-easy-use/index.md

**Reply by snowyu** [reply]
@zu1k 你是保存了两份文件? 一份html,另一份md. 这样有点浪费空间就是了.

我更希望类似这样:

最最基础, for programmers:
```bash
git clone https://zu1k.com/posts/tutorials/p2p
# then some modification...
git pull-request
```

或者能直接在页面上修改后自动产生PR. 这样最好, for common people.

**Comment by bzsome** [comment]
我正在使用ipfs网络恢复您

**Reply by zu1k** [reply]
No, you are not. You are using GitHub Discussion.

**Comment by foool** [comment]
ipfs当初给我“惊喜”除了“基于数据内容寻址”，还有“基于数据价值的文件缓存”/“’付费‘存储”，越有价值的数据被赋值和存储的节点更多，这种方法天然地对垃圾数据和无价值数据进行了筛选。

**Reply by zu1k** [reply]
对，这隐含在IPFS的运作原理之中，这种筛选是潜移默化而又非常必要的

**Reply by livid** [reply]
有时候我在想一些 Legacy 问题——也就是，如果内容创造者死了之后，他的内容怎么办。但可能这件事情的根本取决于内容本身的价值。如果内容足够有意义，那就肯定会有足够多的副本留下来。

**Reply by zu1k** [reply]
很多存世的名人语录都不是本人编纂的，而是徒弟或者其他后人整理并流传下来的，从这个角度看，创作者只需要负责生产好高质量内容即可

**Reply by FledgeXu** [reply]
有价值的内容会被不断的引用、访问和保存，没有价值的内容会被自然的遗忘。
需要解决的不是说如何保存内容，需要解决的是如何让公众可以发现有价值的内容。

**Reply by snowyu** [reply]
发现重要,协作也很重要

**Comment by FledgeXu** [comment]
这里说的更加IPFS的设计其实已经有了，简单来说就是把这个结构设计成一个MST，每次博客文章或者广义上的Post更新，其实就是增加一个Record，这样的话，别人看需不需要更新只需要看最上面的CID有没有变化就行。
具体的设计细节我觉得完全可以参照AT Protocol：https://atproto.com/guides/data-repos

**Comment by Liudon** [comment]
参考大佬文章，搞了套利用github actions部署博客到IPFS，文章见https://liudon.com/posts/deploy-blog-to-ipfs/

[![pSzAWKe.png](https://s1.ax1x.com/2023/02/24/pSzAWKe.png)](https://imgse.com/i/pSzAWKe)

已pin，欢迎大佬也pin下我的地址

`ipfs pin add /ipns/liudon.xyz`

**Reply by livid** [reply]
@liudon 可能还需要处理一下相对链接。现在如果不是在顶级域名里打开就会有问题：

https://ipfs.io/ipns/liudon.xyz

一种简单方式就是：

```
npx all-relative
```

**Reply by Liudon** [reply]
感谢，我用的`hugo`，配置文件里开启`relativeURLs`就可以了，问题已解决。

```
relativeURLs: true
```

**Comment by zu1k** [comment]
发现了一个新平台 xLog，看起来还挺有意思的，也是基于 IPFS 和 区块链钱包来做的，内容放 IPFS 上 ( https://ipfs.io/ipfs/bafkreihem2rblihydfrmzup5rizjqvzptp2ccynhekigm55ddgnogcwtgu )， 发布用 Crossbell 子网来做 ( https://scan.crossbell.io/address/0x8425c55fBd4c90fcD394243195dE388326cC1e42 )

**Reply by urkbio** [reply]
我在用Planet和xlog，但更偏向Planet。xlog没什么好折腾的😅。而且Planet的原版内容是在本地的，尽管xlog也存在IPFS。
还有我看了您的文章，才意识到Planet文章更新渲染机制这样的问题，确实对于IPFS来说不是最优方案。
Planet可以直接关注其他IPFS上的博客，关注了就自动PIN，还会更新。
